amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpGraphStageLogicConnectionShutdownSpec.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp.scaladsl
+
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.ActorSystem
+import akka.dispatch.ExecutionContexts
+import akka.stream.ActorMaterializer
+import akka.stream.alpakka.amqp.{
+  AmqpCachedConnectionProvider,
+  AmqpConnectionFactoryConnectionProvider,
+  AmqpProxyConnection,
+  AmqpSinkSettings,
+  QueueDeclaration
+}
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import com.rabbitmq.client.{AddressResolver, Connection, ConnectionFactory, ShutdownListener}
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.{BeforeAndAfterEach, Matchers, WordSpec}
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+/**
+ * see [[https://github.com/akka/alpakka/issues/883]] and
+ * [[https://github.com/akka/alpakka/pull/887]]
+ */
+class AmqpGraphStageLogicConnectionShutdownSpec
+    extends WordSpec
+    with Matchers
+    with ScalaFutures
+    with BeforeAndAfterEach {
+
+  override implicit val patienceConfig = PatienceConfig(10.seconds)
+  private implicit val executionContext = ExecutionContexts.sameThreadExecutionContext
+
+  val shutdownsAdded = new AtomicInteger()
+  val shutdownsRemoved = new AtomicInteger()
+
+  override def beforeEach() = {
+    shutdownsAdded.set(0)
+    shutdownsRemoved.set(0)
+  }
+
+  trait ShutdownListenerTracking { self: AmqpProxyConnection =>
```

Comment:
A few notes:

I wouldn't have to require `AmqpProxyConnection` as the self-type (`Connection` should be enough), except I couldn't figure out any scala gymnastics to both override `addShutdownListener` and delegate to the super class' default behavior without a stack overflow (infinite loop of this function calling itself). So I require the proxy and get direct access to the delegate.

I could make this trait a bit more general purpose by giving it its own delegate, but it doesn't seem worth the effort to me. Something like this:

```scala
trait ShutdownListenerTracking { self: Connection => 
  protected def delegate: Connection

  override def addShutdownListener(shutdownListener: ShutdownListner): Unit = {
    shutdownsAdded.incrementAndGet
    delegate.addShutdownListener(shutdownListener)
  }
}
```